Add content for local tools (#17065)
dotnet/docs: docs/core/tools/global-tools-how-to-create.md

---
title: How to create a .NET Core tool
description: Learn how to create a .NET Core tool. A tool is a console application that is installed by using the .NET Core CLI.
author: tdykstra
ms.author: tdykstra
ms.date: 02/12/2020
---

# Create a .NET Core tool using the .NET Core CLI

**This article applies to:** ✔️ .NET Core 3.x SDK

This tutorial teaches you how to create and package a .NET Core tool. The .NET Core CLI lets you create a console application as a tool, which others can install and run. .NET Core tools are NuGet packages that are installed from the .NET Core CLI. For more information about tools, see [.NET Core tools overview](global-tools.md).

The tool that you'll create is a console application that takes a message as input and displays the message along with lines of text that create the image of a robot.

This is the first in a series of three tutorials. In this tutorial, you create and package a tool. In the next two tutorials you [use the tool as a global tool](global-tools-how-to-use.md) and [use the tool as a local tool](local-tools-how-to-use.md).

## Prerequisites

> - [.NET Core SDK 3.1](https://dotnet.microsoft.com/download) or a later versions.

Reviewer: ```suggestion
- [.NET Core SDK 3.1](https://dotnet.microsoft.com/download) or a later version.
```

- A text editor or code editor of your choice.

## Create a project

1. Open a command prompt and create a folder named *botsay*.

> 1. Navigate to the folder you created and type the following:

Reviewer: ```suggestion
1. Navigate to the folder you created and enter the following command:
```

More accessibility friendly to say enter instead of type.

   ```dotnetcli
   dotnet new console
   ```

## Add the code

1. Open the `Program.cs` file with your code editor.

1. Add the following `using` directive to the top of the file:

   ```csharp
   using System.Reflection;
   ```

> 1. Replace the `Main` method with the following code to process the command-line arguments for the application. If no arguments were passed, a short help message is displayed. Otherwise, all of the arguments are concatenated into a single string and printed by calling the `ShowBot` method. You create the `ShowBot` method in the next step.

Reviewer: ```suggestion
1. Replace the `Main` method with the following code to process the command-line arguments for the application. If no arguments are passed, a short help message is displayed. Otherwise, all of the arguments are concatenated into a single string and printed by calling the `ShowBot` method. You create the `ShowBot` method in the next step.
```

Also consider moving the description of the code below the actual code.

1. Open the `botsay.csproj` file and add three new XML nodes to the end of the `<PropertyGroup>` node:

   ```xml
   <PackAsTool>true</PackAsTool>
   <ToolCommandName>botsay</ToolCommandName>
   <PackageOutputPath>./nupkg</PackageOutputPath>
   ```

   `<ToolCommandName>` is an optional element that specifies the command that will invoke the tool after it's installed. If this element isn't provided, the command name for the tool is the project file name without the *.csproj* extension.

   `<PackageOutputPath>` is an optional element that determines where the NuGet package will be produced. The NuGet package is what the .NET Core CLI uses to install your tool.

   The project file now looks like the following example:

   ```xml
   <Project Sdk="Microsoft.NET.Sdk">
  
     <PropertyGroup>

       <OutputType>Exe</OutputType>
       <TargetFramework>netcoreapp3.1</TargetFramework>
  
       <PackAsTool>true</PackAsTool>
       <ToolCommandName>botsay</ToolCommandName>
       <PackageOutputPath>./nupkg</PackageOutputPath>
  
     </PropertyGroup>

   </Project>
   ```

1. Create a NuGet package by running the [dotnet pack](dotnet-pack.md) command:

>   ```dotnetcli

Reviewer: ```suggestion
   ```dotnetcli
```

>   dotnet pack

Reviewer: ```suggestion
   dotnet pack
```

>   ```

Reviewer: ```suggestion
   ```
```

>   The `botsay.1.0.0.nupkg` file is created in the folder identified by the `<PackageOutputPath>` value from the `botsay.csproj` file, which in this example is the `./nupkg` folder.

Reviewer: ```suggestion
   The `botsay.1.0.0.nupkg` file is created in the folder identified by the `<PackageOutputPath>` value from the `botsay.csproj` file, which in this example is the `./nupkg` folder.
```

  
  When you want to release a tool publicly, you can upload it to <https://www.nuget.org>. Once the tool is available on NuGet, developers can install the tool by using the [dotnet tool install](dotnet-tool-install.md) command.

## Troubleshooting

> If you get an error message while following the tutorial, see [Troubleshoot .NET Core tool usage issues](troubleshoot-usage-issues.md)

Reviewer: ```suggestion
If you get an error message while following the tutorial, see [Troubleshoot .NET Core tool usage issues](troubleshoot-usage-issues.md).
```
